framework/src/media/MediaRecorderImpl.cpp

```diff
@@ -171,37 +176,58 @@ recorder_result_t MediaRecorderImpl::stop()
 
 recorder_result_t MediaRecorderImpl::pause()
 {
-	lock_guard<mutex> lock(mCmdMtx);
+	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaRecorderImpl::pause()\n");
 	RecorderWorker& mrw = RecorderWorker::getWorker();
 	mrw.getQueue().enQueue(&RecorderWorker::pauseRecorder, &mrw, shared_from_this());
 
 	return RECORDER_OK;
 }
 
-recorder_result_t MediaRecorderImpl::getVolume()
+int MediaRecorderImpl::getVolume()
 {
-	lock_guard<mutex> lock(mCmdMtx);
+	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaRecorderImpl::getVolume()\n");
-	return RECORDER_OK;
+
+	if (mCurState == RECORDER_STATE_IDLE || mCurState == RECORDER_STATE_NONE) {
+		medvdbg("MediaRecorderImpl::getVolume() - mCurState == RECORDER_STATE_IDLE || mCurState == RECORDER_STATE_NONE\n");
+		return 0;
+	}
+
+	return mCurVolume;
 }
 
-void MediaRecorderImpl::setVolume(int vol)
+recorder_result_t MediaRecorderImpl::setVolume(int vol)
 {
-	lock_guard<mutex> lock(mCmdMtx);
+	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaRecorderImpl::setVolume(int vol)\n");
+
+	if (mCurState == RECORDER_STATE_IDLE || mCurState == RECORDER_STATE_NONE) {
+		medvdbg("MediaRecorderImpl::setVolume(int vol) - mCurState == RECORDER_STATE_IDLE || mCurState == RECORDER_STATE_NONE\n");
+		return RECORDER_ERROR;
+	}
+
 	mCurVolume = vol;
+	return RECORDER_OK;
 }
 
-void MediaRecorderImpl::setDataSource(std::unique_ptr<stream::OutputDataSource> dataSource)
+recorder_result_t MediaRecorderImpl::setDataSource(std::unique_ptr<stream::OutputDataSource> dataSource)
 {
-	lock_guard<mutex> lock(mCmdMtx);
-	medvdbg("MediaRecorderImpl::setDataSource(OutputDataSource *_OutputDataSource)\n");
-	mOutputDataSource = std::move(dataSource);
+	std::lock_guard<std::mutex> lock(mCmdMtx);
+	medvdbg("MediaRecorderImpl::setDataSource(std::unique_ptr<stream::OutputDataSource> dataSource)\n");
+	
+	if (mCurState != RECORDER_STATE_IDLE || !dataSource) {
```

Comment:
Let's seperate this condition to debug more easily.
They are diffrent type condition, I think

Reply:
i think, not required seperate.

Reply:
If other error codes are added, separation is necessary, but it is not necessary at present.